store/src/qty.rs

```diff
@@ -0,0 +1,798 @@
+    let mut data = data.clone();
+
+    let mut root = Qty { number: data["number"].number(), name: Uom::In(UUID_NIL, None) };
+
+    let mut head = &mut root;
+
+    while data.is_object() {
+      let uom = data["uom"].clone();
+      let mut tmp = Qty { number: Decimal::ZERO, name: Uom::In(UUID_NIL, None) };
+      if uom.is_object() {
+        tmp.number = uom["number"].number();
+        tmp.name = Uom::In(UUID_NIL, None);
+        head.name = Uom::In(uom["in"].uuid()?, Some(Box::new(tmp.clone())));
+        head = match &mut head.name {
+          Uom::In(_, ref mut qty) => qty.as_mut().unwrap(),
+          _ => unreachable!(),
+        };
+      } else {
+        head.name = Uom::In(uom.uuid()?, None);
+      }
+      data = uom;
+    }
+
+    Ok(root)
+  }
+
+  fn from_vec(data: Vec<(Decimal, Uuid)>) -> Result<Self, WHError> {
+    if data.len() < 1 {
+      return Err(WHError::new("empty vector"));
+    }
+    let mut data = data.clone();
+    data.reverse();
+
+    let mut root = Qty { number: data[0].0, name: Uom::In(data[0].1, None) };
+
+    let mut head = &mut root;
+
+    for (num, id) in data[1..].iter() {
+      let tmp = Qty { number: num.clone(), name: Uom::In(id.clone(), None) };
+      head.name = Uom::In(head.name.uuid(), Some(Box::new(tmp)));
+      head = match &mut head.name {
+        Uom::In(_, ref mut qty) => qty.as_mut().unwrap(),
+      };
+    }
+
+    Ok(root)
+  }
+
+  fn qty(&self) -> Option<Box<Qty>> {
+    match &self.name {
+      Uom::In(_, qty) => qty.clone(),
+    }
+  }
+
+  fn depth(&self) -> Vec<(Decimal, Uuid)> {
+    let mut result = Vec::new();
+
+    result.insert(0, (self.number, self.name.uuid()));
+
+    let mut current = self.clone();
+
+    while let Some(qty) = current.qty() {
+      result.insert(0, (qty.number, qty.name.uuid()));
+      current = qty.as_ref().clone();
+    }
+
+    result
+  }
+
+  fn simplify(data: &mut Vec<(Decimal, Uuid)>, index: usize) {
+    // println!("simplify: {data:?}");
+    while data.len() > index + 1 {
+      if let (Some(popped), Some(last)) = (data.pop(), data.last_mut()) {
+        let (last_number, last_uuid) = last.clone();
+        *last = (popped.0 * last_number, last_uuid);
+      }
+    }
+  }
+
+  fn subtract(
+    full: Vec<(Decimal, Uuid)>,
+    mut left: Vec<(Decimal, Uuid)>,
+    right: Vec<(Decimal, Uuid)>,
+  ) -> Result<Vec<Self>, WHError> {
+    if let (Some(l_last), Some(r_last)) = (left.last_mut(), right.last()) {
+      let (last_number, last_uuid) = l_last.clone();
+      let number = last_number - r_last.clone().0;
+      if number == Decimal::ZERO {
+        return Ok(Vec::new());
+      }
+      *l_last = (number, last_uuid);
+    }
+
+    left.convert(full)
+  }
+}
+
+impl Add for Qty {
+  type Output = Vec<Self>;
+
+  fn add(self, rhs: Self) -> Self::Output {
+    let mut result = self.clone();
+    let mut vector: Vec<Self> = Vec::new();
+
+    let mut left = Uom::In(self.name.uuid(), Some(Box::new(self.clone())));
```

Comment:
why `cloned`?

Reply:
Because I need these objects later in the method. I added an if statement and put it inside. Cloning will not happen without need anymore.